### DISCUSSION TITLE
Output directory for tar_quarto_rep()

### DISCUSSION THREAD

**Discussion by mglev1n** [ropensci/tarchetypes#129] [Ideas]
Currently, when using `tar_quarto_rep()` the rendered quarto output (eg. *.html, *.pdf) seem to be placed in the main targets working directory, regardless of any path provided using the `output_file` parameter in `execute_params`. This is related to https://github.com/ropensci/tarchetypes/discussions/107, but I think different enough to warrant separate discussion:

```
lines <- c(
  "---",
  "title: 'report.qmd file'",
  "output_format: html_document",
  "params:",
  "  par: \"default value\"",
  "---",
  "Assume these lines are in a file called report.qmd.",
  "```{r}",
  "print(params$par)",
  "```"
)
writeLines(lines, "report.qmd") # In tar_dir(), not the user's file space.
dir.create("results")

targets::tar_script({
  library(tarchetypes)
  list(
    tar_quarto_rep(
      report,
      path = "report.qmd",
      # reports are placed in working directory, rather than "Results/"
      execute_params = tibble::tibble(
        par = c(1, 2),
        output_file = c("results/rendered_report1.html", "results/rendered_report2.html")
      )
    )
  )
}, ask = FALSE)
```
This behavior quickly clutters the working directory as the number of reports increases. It seems like the report must be rendered from the working directory in order to maintain access to the target store, however once rendering is complete the output no longer seems dependent on the working directory. The most straightforward solution is to manually move the rendered reports to a new directory for storage, but this defeats the purpose of using targets. I'm curious if a solution may be to provide a `storage_directory` or `output_directory` parameter or argument to the `tar_quarto` and `tar_render` family of functions. Once rendered, the reports could be moved to this location (Eg. `fs::file_move()` or `file.rename()` prior to being hashed/logged in storage?

I realize this is an edge-case, but am curious if there may be other solutions? Presumably I could wrap `quarto::quarto_render()` and a function to move the rendered document?

**Comment by wlandau** [comment]
Yeah, you're right. Thanks for the nice example. e934e5d285f8f2a56f5cc87277a736211fa0cb1c should fix it.

**Reply by benzipperer** [reply]
I'm using tarchetypes 0.7.4 but can't get tar_quarto_rep() to run when there are paths in the output_file values. In particular, the example above gives me the quarto error

```r
targets::tar_make()
• start target report_params
• built target report_params [0.004 seconds]
• start branch report_056eda16
✖ error branch report_056eda16
• end pipeline [0.238 seconds]
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: ! ! System command 'quarto' failed
```

It runs fine when I remove the "Results" path. The directory "Results" exists and is writeable but empty.

Does the example from @mglev1n above work for you?  Do you have a suggestion how I can narrow down the problem?

Thank you for the excellent packages!

**Reply by wlandau** [reply]
I tried the reprex again, but it worked on my end when I created the `Results` directory in advance using `dir.create("Results")`. So in your case it could be an issue with Quarto. You can get more specific error messages if you put `quiet = TRUE` in `tar_quarto_rep()`. https://github.com/quarto-dev/quarto-r/issues/16 is still an open issue.

**Reply by benzipperer** [reply]
interesting, so when I try `quiet = FALSE` in `tar_quarto_rep()` I get an error suggesting the path is passed to the `--output` option of `quarto render` which apparently only accepts a filename? 

```r
✔ skip target report_params
• start branch report_056eda16
ERROR: --output option cannot specify a relative or absolute path
✖ error branch report_056eda16
• end pipeline [0.229 seconds]
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: ! ! System command 'quarto' failed
```

I'm a little confused about what's driving the error here and why it seems to work for you. I'm using quarto cli 1.2.335 and quarto R 1.2.

**Reply by benzipperer** [reply]
I wonder if the issue is that you're using an older version of quarto cli? it looks like commit [bd6de52](https://github.com/quarto-dev/quarto-cli/commit/bd6de5295e482b805739e8b76faf8ab622c13cc3) disallows paths in --output

**Reply by wlandau** [reply]
After upgrading Quarto CLI, I see exactly what you see. I am surprised Quarto decided to go that direction. For now, I pushed a patch that makes the example work: 68e9ea1d210e8998a4ef04a6d0b9d817141cc10c
